src/PhpParser/Node/BetterNodeFinder.php

```diff
@@ -452,41 +431,6 @@ public function resolveCurrentStatement(Node $node): ?Stmt
         return null;
     }
 
-    /**
-     * @api
-     *
-     * Resolve previous node from any Node, eg: Expr, Identifier, Name, etc
-     */
-    public function resolvePreviousNode(Node $node): ?Node
```

Comment:
This is one that I prefer to keep, this resolve previous node from any Node except Stmt, eg: from Identifier or from Name or Expr, there is no way to get previous node without resolve current stmt, eg: used on `RemoveNonExistingVarAnnotationRector` from `Assign` node.

https://github.com/rectorphp/rector-src/blob/21f8014b30f4a4230f787ca1698a2c53d6d5f18a/rules/DeadCode/Rector/Node/RemoveNonExistingVarAnnotationRector.php#L43-L44


Also, with make it public means it help user refactor their code on migration to 1.0

Reply:
We are removing all the previous/next/parent finding from BetterNodeFinder. Rector 1.0 will not have them at all. The method is currently unused as well.

Reply:
I prefer to make "parent" as "internal" tho, as we know the scope from there. That we recommend to not use in their code, while it can with cautions above performance impact.

This change also make `findFirstInlinedPrevious()` no longer correct, see my comment below https://github.com/rectorphp/rector-src/pull/4192/files#r1226497153

Reply:
The goal is to remove them even for internals, as they take up the memory for each parent node. That's why all these methods will not work and have to removed and changed beforehand.

Like PHPStan did last year: https://phpstan.org/blog/preprocessing-ast-for-custom-rules

Now we force PHPStan to parse nodes with parent node in each single node, so any memory savings mentioned in the post are gone. The goal is to avoid any parent node traversal anywhere, and make the node changes explicit from top to bottom path.

Reply:
All the findFirstPrevious that read from Expr, Name, or Identifier need to be migrated to read from Stmt before it can actually be removed, eg, on PR:

- https://github.com/rectorphp/rector-src/pull/4209